### PR TITLE
Cookie-based client sessions with specific scopes and tenants

### LINKED ISSUE #127: Fix cookie-based client sessions
## Current behavior
- When `/oidc/authorize` is requested with `cookie` in scope, a new client session is created. However, this session is then never used, as the cookie always directly identifies the root session. Each such call to `authorize` always only creates a new client session.

## Desired behavior
- Cookie introspection is always configured for a specific client_id. Session is matched by the combination of SCI and client_id, which should be unique.
- This should also remove the need for cookie domain being part of the cookie value. The domain should be specified in the client configuration.

## Implementation

- [ ] Create compound index `(cookie_sci, client_id)` in the session collection 
- [ ] Remove the domain prefix from the cookie value
- [ ] Add mandatory `client_id` parameter to cookie introspection endpoint
- [ ] Cookie domain is extracted from client metadata (`client_uri` or a new dedicated attribute)
- [ ] Authorize call with `scope=cookie` responds with Set-Cookie header configured for the respective client domain.

### PR DESCRIPTION
**Pull request by byewokko**
closes #127 

# Changes
- **`BREAKING`** Authorize call now rejects unknown `client_id`s.
  - :warning:  All client applications must be registered via client registration endpoint.
- Cookie introspection call **with `client_id`** in query string authenticates with corresponding **client** session. The `client_id` must be a valid registered client.
- Cookie introspection call **without `client_id`** in query string authenticates with corresponding **root** session
- Created secondary indexes in session collection (access token, cookie+client_id)
- Refactored `SessionService.get_by` method to allow searching by compound indexes.
- Standardized error codes in authorize response (see [RFC6749](https://www.ietf.org/rfc/rfc6749.txt) section 4.1.2.1 and [OIDC Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html#AuthError))

# Example usage

- First, register your `demo-app` Public web app client either using the client registration form in the Admin UI or using the API `POST /client/register`.
- Set up the introspection endpoint with the `demo-app` client ID in the query:

```nginx
location = /_demo_cookie_introspect {
	internal;
	proxy_method         POST;
	proxy_pass           http://auth_api/cookie/nginx?client_id=demo-app;
	proxy_set_body       "$http_authorization";
	proxy_ignore_headers Cache-Control Expires Set-Cookie;
}
```

- Use the introspection `auth_request` in your protected location. You can add the `error_page` directive to redirect unauthorized users to the authorize endpoint to obtain the cookie.

```nginx
location /demo {
	auth_request /_demo_cookie_introspect;
	error_page 401 /auth/api/openidconnect/authorize?client_id=demo-app&response_type=code&scope=openid%20cookie%20tenant&redirect_uri=$request_uri;

	auth_request_set      $authorization $upstream_http_authorization;
	proxy_set_header      Authorization $authorization;

	auth_request_set      $cookie $upstream_http_cookie;
	proxy_set_header      Cookie $cookie;

	auth_request_set   $set_cookie $upstream_http_set_cookie;
	add_header	Set-Cookie $set_cookie;
	rewrite ^/demo(/(.*))? /$2 break;
	proxy_pass http://localhost:8080;
}
```

### CODE REVIEW

File: seacatauth/authz/rbac/handler.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import asab.web.rest
 
 from ...decorators import access_control
-from ...exceptions import TenantRequired
+from ...exceptions import TenantNotSpecified
```
Comment (ateska) [inline seacatauth/authz/rbac/handler.py:6]: Exceptions should end with `Error` -> `TenantNotSpecifiedError`

**Review by ateska — CHANGES_REQUESTED** [review]
This is a long merge request.
Some feedback provided but it is for a workshop to merge it.

File: seacatauth/openidconnect/handler/authorize.py [hunk 2]
```diff
@@ -161,12 +162,22 @@ async def authentication_code_flow(
...
+			return self.reply_with_authentication_error(
+				"invalid_client_id",
+				redirect_uri,
+				error_description="Invalid client_id",
+				state=state
+			)
+		except client_exceptions.InvalidClientSecret as e:
+			L.warning(str(e), struct_data={"client_id": client_id})
+			return self.reply_with_authentication_error(
+				"unauthorized_client",
```
Comment (ateska) [inline seacatauth/openidconnect/handler/authorize.py:175]: Please add info that this value is from a standard (link to the specs)

Comment (byewokko) [reply seacatauth/openidconnect/handler/authorize.py]: I extracted these error type strings into the member variables of `openidconnect.utils.AuthErrorResponseCode` and supplemented links to corresponding RFC standard.

**Pull request merged**